lib/ide/parser/src/api.rs

```diff
@@ -1,20 +1,57 @@
 use crate::prelude::*;
+use ast::ID;
+use std::ops::Sub;
 
-pub type Ast = ast::Ast;
+use serde::Serialize;
+use serde::Deserialize;
 
+pub type Ast = ast::Ast;
 
 
 // ============
 // == Parser ==
 // ============
 
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Index { pub value:usize }
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Size { pub value:usize }
+
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct Span { pub index: Index, pub size: Size }
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub struct IDMap(pub Vec<(Span, ID)>);
+
+impl Add for Size {
+    type Output = Size;
+    fn add(self, rhs:Size) -> Size {
+        Size { value: self.value + rhs.value }
+    }
+}
+
+impl Add<Size> for Index {
+    type Output = Index;
+    fn add(self, rhs:Size) -> Index {
+        Index { value: self.value + rhs.value }
+    }
+}
+
+impl Sub<Size> for Index {
+    type Output = Index;
+    fn sub(self, rhs:Size) -> Index {
+        Index { value: self.value - rhs.value }
```

Comment:
theres should be no spaces inside {…}  (here and same applies to instances above)